File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/AnnotationRule.java
package com.omegar.mvp.compiler;

import com.omegar.mvp.compiler.entity.AnnotationInfo;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;

/**
 * Date: 17-Feb-16
 * Time: 16:57
 *
 * @author esorokin
 */
public abstract class AnnotationRule {
	protected final Set<Modifier> mValidModifiers;
	protected final AnnotationInfo<?> mAnnotationInfo;
	protected StringBuilder mErrorBuilder;

	public AnnotationRule(AnnotationInfo<?> annotationInfo, Modifier... validModifiers) {
		mAnnotationInfo = annotationInfo;
		if (validModifiers == null || validModifiers.length == 0) {
			throw new RuntimeException("Valid modifiers cant be empty or null.");
		}

		mValidModifiers = new HashSet<>(Arrays.asList(validModifiers));
		mErrorBuilder = new StringBuilder();
	}

	/**
	 * Method describe rules for using Annotation.
	 *
	 * @param AnnotatedField Checking annotated field.
	 */
	public abstract void checkAnnotation(Element AnnotatedField);

	public String getErrorStack() {
		return mErrorBuilder.toString();
	}

	protected String validModifiersToString() {
		if (mValidModifiers.size() > 1) {
			StringBuilder result = new StringBuilder("one of [");
			boolean addSeparator = false;
			for (Modifier validModifier : mValidModifiers) {
				if (addSeparator) {
					result.append(", ");
				}
				addSeparator = true;
				result.append(validModifier.toString());
			}
			result.append("]");
			return result.toString();
		} else {
			return mValidModifiers.iterator().next() + ".";
		}
	}
}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/AnnotationRule.kt
package com.omegar.mvp.compiler

import com.omegar.mvp.compiler.entity.AnnotationInfo
import javax.lang.model.element.Element
import javax.lang.model.element.Modifier

/**
 * Date: 17-Feb-16
 * Time: 16:57
 *
 * @author esorokin
 */
abstract class AnnotationRule(val annotationInfo: AnnotationInfo<*>, vararg validModifiers: Modifier) {

    protected val validModifiers: Set<Modifier> = validModifiers.toSet()

    protected var errorBuilder: StringBuilder = StringBuilder()

    val errorStack: String
        get() = errorBuilder.toString()

    init {
        if (validModifiers.isEmpty()) {
            throw RuntimeException("Valid modifiers cant be empty or null.")
        }
    }

    /**
     * Method describe rules for using Annotation.
     *
     * @param annotatedField Checking annotated field.
     */
    abstract fun checkAnnotation(annotatedField: Element)

    protected fun validModifiersToString(): String {
        return if (validModifiers.size > 1) {
            val result = StringBuilder("one of [")
            var addSeparator = false
            for (validModifier in validModifiers) {
                if (addSeparator) {
                    result.append(", ")
                }
                addSeparator = true
                result.append(validModifier.toString())
            }
            result.append("]")
            result.toString()
        } else {
            validModifiers.iterator().next().toString() + "."
        }
    }

}